libmscore/scale.cpp

```diff
@@ -0,0 +1,265 @@
+                        }
+
+                  originalNotes.push_back(1200 * log(values[0].toDouble() / values[1].toDouble()) / log(2));
+            } else {
+                  originalNotes.push_back(pitchLine.toDouble());
+                  }
+            }
+
+      computedNotes.clear();
+      switch (originalNotes.size()) {
+            case 7:
+                  computeNotes7();
+                  break;
+            case 12:
+                  computeNotes12();
+                  break;
+            case 17:
+                  computeNotes17();
+                  break;
+            case 21:
+                  computeNotes21();
+                  break;
+            case 35:
+                  computeNotes35();
+                  break;
+            default:
+                  MScore::lastError = QObject::tr("Unsupported Scala file format.");
+                  return false;
+            }
+      return true;
+      }
+
+//---------------------------------------------------------
+//   computeNotes8
+//---------------------------------------------------------
+void Scale::computeNotes7() {
+      static vector<Tpc> notesToFill = {TPC_D, TPC_E, TPC_F, TPC_G, TPC_A, TPC_B, TPC_C};
+
+      for (size_t i = 0; i < originalNotes.size(); ++i) {
+            computedNotes[notesToFill[i]] = originalNotes[i];
+            }
+
+      fillMissing7();
+      }
+
+//---------------------------------------------------------
+//   computeNotes12
+//---------------------------------------------------------
+void Scale::computeNotes12() {
+      static vector<Tpc> notesToFill = {
+            TPC_C_S, TPC_D,
+            TPC_D_S, TPC_E,
+                     TPC_F,
+            TPC_F_S, TPC_G,
+            TPC_G_S, TPC_A,
+            TPC_A_S, TPC_B,
+                     TPC_C};
+
+      for (size_t i = 0; i < originalNotes.size(); ++i) {
+            computedNotes[notesToFill[i]] = originalNotes[i];
+            }
+
+      fillMissing12();
+      }
+
+//---------------------------------------------------------
+//   computeNotes17
+//---------------------------------------------------------
+void Scale::computeNotes17() {
+      static vector<Tpc> notesToFill = {
+            TPC_C_S, TPC_D_B, TPC_D,
+            TPC_D_S, TPC_E_B, TPC_E,
+                              TPC_F,
+            TPC_F_S, TPC_G_B, TPC_G,
+            TPC_G_S, TPC_A_B, TPC_A,
+            TPC_A_S, TPC_B_B, TPC_B,
+                              TPC_C};
+
+      for (size_t i = 0; i < originalNotes.size(); ++i) {
+            computedNotes[notesToFill[i]] = originalNotes[i];
+            }
+
+      fillMissing17();
+      }
+
+//---------------------------------------------------------
+//   computeNotes21
+//---------------------------------------------------------
+void Scale::computeNotes21() {
+      static vector<Tpc> notesToFill = {
+            TPC_C_S, TPC_D_B, TPC_D,
+            TPC_D_S, TPC_E_B, TPC_E,
+            TPC_E_S, TPC_F_B, TPC_F,
```

Comment:
Not sure. I would expect Fb (usually near to E) to come before E# (usually near to F).